### ISSUE TITLE
Suggestion for passing username as a Principal from the AWS Lambda context

### ISSUE THREAD

**Issue by richiethom** [quarkusio/quarkus#16647]
## Description

At the moment to get the Cognito account associated with an AWS Lambda HTTP invocation, it is necessary to do the following, which isn't particularly clean:

```java
myAPIGatewayV2HTTPEvent.getRequestContext().getAuthorizer().getJwt().getClaims().get("cognito:username")
```

(this is using the newer AWS API Gateway which doesn't connect with Cognito as easily, we have to integrate Cognito using JWT)

In addition, it doesn't seem possible to inject this information when the Lambda is being run locally in AWS SAM, where there isn't an AWS API Gateway running.

## Implementation ideas

I have made a preliminary implementation of the following method signature in `LambdaHttpHandler`. When run in **native** code, it will return an `Optional<Principal>` containing the `cognito:username` if set. When run in SAM, it will read a username from an environment variable, if another environment variable is also set to true.

```java
private Optional<Principal> getPrincipal(APIGatewayV2HTTPEvent request)
```

If the Principal is present, it is then added to the `quarkusHeaders` object, meaning it is injected into a lambda thus:

```java
@POST
public MyResponse processMyRequest(@javax.ws.rs.core.Context Principal principal, ...)
```

The implementation of the method is in [this](https://github.com/richiethom/quarkus/commit/1fa2cda26f2c5e6cd7e5f8f61554e36c391a510b) branch on my fork. Let me know what you think, and please get back to me with suggestions!

Is there a suitable object for injecting JWT claims in the same way? Perhaps a class that implements Principal but can be cast to allow the JWT claims to be retrieved is a possibility?

**Comment by patriot1burke** [comment]
This is a good workaround for now.   Additionally we'd need to make this Principal available using standard mechanisms for obtaining it (i.e. the SecurityContext interface JAX-RS has).

**Comment by richiethom** [comment]
@patriot1burke Just to be sure I've understood correctly, you're happy for me to simply expose the Principal in the context for the time being, as a workaround, and then a better implementation is to provide a SecurityContext implementation?

I've looked into doing a SecurityContext implementation straight away but it means introducing a JAX-RS dependency into the project. We could mark it as provided but then we'd need to check if the interface was there, which I assume is something to avoid when when running natively.

If you're happy that I inject a simple Principal instance, I'll sort out the pull request.

**Comment by patriot1burke** [comment]
Yes, injecting a simple principal instance is fine for now.

**Comment by jonathanroques** [comment]
Will this "patch" be merged in the next released ? Indeed it would be something very useful for both tests and production.

Thanks

**Comment by richiethom** [comment]
Hi @jonathanroques I'll try and work on a pull request in the next few days. Once the pull request is done, don't hesitate to let me know what you think. One aspect that you might have an opinion on: do you see a need for the production injection of Principal to happen in a non-native environment? I can't see someone wanting to run interpreted code in production, one major benefit of Quarkus is being able to compile to native of course.

**Comment by jonathanroques** [comment]
Agreed indeed, but the builds (native and jvm) should always be _iso-featured_ right ? That would be the only reason I guess.

**Comment by richiethom** [comment]
@patriot1burke , @jonathanroques I have made an initial pull request, which I'd appreciate your review and input on. In particular:

- should this feature only be available when not running natively, as a security feature? I'm in two minds. I can see @jonathanroques' point of view about being iso-featured, but can also see it as a failsafe
- are the variable names ok?
- I have added the Environment section to the sam.jvm.yaml file. Should it be added to the native environment too? Kind of depends on my first question.
- Should the code detect when the `QUARKUS_AWS_LAMBDA_FORCE_USER` flag has been set to `true` but the `QUARKUS_AWS_LAMBDA_FORCE_USER_NAME` variable hasn't been changed? (As I understand it, for a variable to be settable in SAM, it needs to be declared with a value in the initial template file, even if it is overridden with a parameters file)

I'm sure there are things I've missed. Let me know what you think.

**Comment by patriot1burke** [comment]
Sorry, I missed the part about native and jvm mode being different.   I'm not understanding why native and JVM mode are different.

**Comment by patriot1burke** [comment]
This is not 'native mode' vs. 'jvm mode', right?  You have the environment variable in place for SAM LOCAL testing only, correct?  Not sure if I like the idea of modifying the generated .yaml files to include setting this ENVIRONMENT variable.  Can't we just document this feature?  Or just generate separate testing .yaml files?

**Comment by sberyozkin** [comment]
@jonathanroques 

> Is there a suitable object for injecting JWT claims in the same way? Perhaps a class that implements Principal but can be cast to allow the JWT claims to be retrieved is a possibility?

Yes, `quarkus-smallrye-jwt` provides a `JsonWebToken` which also implements `Principal`

**Comment by jonathanroques** [comment]
PR merged I guess you can close this issue :)

**Comment by richiethom** [comment]
@patriot1burke any chance this could be backported into a hypothetical 1.13.4 release if one is planned?
